### PR TITLE
integ test

### PR DESCRIPTION
**Pull request by DIBL**

### CODE REVIEW

File: src/test/com/Elessar/app/client/MyClientIntegTest.java [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+        try {
+            try (MongoCursor<Document> cursor = mongoDB.getCollection(MyDatabase.USERS).find(or(eq(User.NAME, "Shuai Hu"),
+                    eq(User.NAME, "Zhi Xu"))).iterator()) {
+                assertEquals(false, cursor.hasNext());
+            }
+
+            RegistrationResponse registerResp1 = clientA.register("Shuai Hu", "4238902", "shuaih@yahoo.com", "62357492");
+            assertEquals(true, registerResp1.getSuccess());
+
+            RegistrationResponse registerResp2 = clientA.register("Shuai Hu", "4238902", "shuaih@yahoo.com", "62357492");
```
Comment (hypnotichen) [inline src/test/com/Elessar/app/client/MyClientIntegTest.java]: you should add another failed case where username is the same but password is different.

File: src/test/com/Elessar/app/client/MyClientIntegTest.java [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+            assertEquals(false, registerResp2.getSuccess());
+
+            RegistrationResponse registerResp3 = clientA.register("Zhi Xu", "213252", "zhix@126.com", "62258237");
+            assertEquals(true, registerResp3.getSuccess());
+
+            try (MongoCursor<Document> cursor = mongoDB.getCollection(MyDatabase.USERS).find(eq(User.NAME, "Shuai Hu")).iterator()) {
+                assertEquals(true, cursor.hasNext());
```
Comment (hypnotichen) [inline src/test/com/Elessar/app/client/MyClientIntegTest.java]: Also check:
- total number of documents returned is 1
- password, email and phone number are the same as requested

Same for the other user

File: src/test/com/Elessar/app/client/MyClientIntegTest.java [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+            System.out.println(e.getMessage());
+            assert(false);
+        }
+    }
+
+
+    private void logOnTest() {
+        try {
+            try (MongoCursor<Document> cursor = mongoDB.getCollection(MyDatabase.USERS).find(eq(User.NAME, "Shuai Hu")).iterator()) {
+                if (cursor.hasNext()) {
```
Comment (hypnotichen) [inline src/test/com/Elessar/app/client/MyClientIntegTest.java]: Probably better to cleanup the collection and do the setup again here.
We can have 50 private methods testing different things and someone might add a new test method in the middle and accidentally breaks all tests after it.

Here if registrationTest() fails and Users table doesn't have this user we won't even notice something is wrong as we are only doing the assertion inside the if.

Same for the other tests

File: src/test/com/Elessar/app/client/MyClientIntegTest.java [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+            Assert.assertEquals(false, logonResp2.getSuccess());
+
+            LogonResponse logonResp3 = clientA.logOn("Shuai Hu", "4238902", clientA_Port);
+            Assert.assertEquals(true, logonResp3.getSuccess());
+
+            LogonResponse logonResp4 = clientA.logOn("Shuai Hu", "4238902", clientA_Port);
+            Assert.assertEquals(true, logonResp4.getSuccess());
+
+            try (MongoCursor<Document> cursor = mongoDB.getCollection(MyDatabase.USERS).find(eq(User.NAME, "Shuai Hu")).iterator()) {
+                if (cursor.hasNext()) {
```
Comment (hypnotichen) [inline src/test/com/Elessar/app/client/MyClientIntegTest.java]: Should assert cursor.hasNext() or total number of documents returned is 1

File: src/test/com/Elessar/app/client/MyClientIntegTest.java [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+                    assertEquals(true, cursor.next().getBoolean(User.ONLINE));
+                }
+            }
+
+            LogoffResponse logoffResp1 = clientA.logOff("Ertai Cai");
+            Assert.assertEquals(false, logoffResp1.getSuccess());
+
+            LogoffResponse logoffResp2 = clientA.logOff("Zhi Xu");
```
Comment (hypnotichen) [inline src/test/com/Elessar/app/client/MyClientIntegTest.java]: Better add a comment here and also for logoffResp3 mentioning what we are testing here: e.g. one is when user registered but offline and the other is user registered and online.

File: src/test/com/Elessar/app/client/MyClientIntegTest.java [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+            Assert.assertEquals(true, logoffResp2.getSuccess());
+
+            LogoffResponse logoffResp3 = clientA.logOff("Shuai Hu");
+            Assert.assertEquals(true, logoffResp3.getSuccess());
+
+            LogoffResponse logoffResp4 = clientA.logOff("Shuai Hu");
+            Assert.assertEquals(true, logoffResp4.getSuccess());
+
+            try (MongoCursor<Document> cursor = mongoDB.getCollection(MyDatabase.USERS).find(eq(User.NAME, "Shuai Hu")).iterator()) {
+                if (cursor.hasNext()) {
```
Comment (hypnotichen) [inline src/test/com/Elessar/app/client/MyClientIntegTest.java]: same comment for the "if" here: we should get rid of this if.
also should do the same check for user "Zhi Xu"

File: src/test/com/Elessar/app/client/MyClientIntegTest.java [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+
+            LogonResponse logonResp1 = clientA.logOn("Shuai Hu", "4238902", clientA_Port);;
+            assertEquals(true, logonResp1.getSuccess());
+
+            P2PMsgResponse p2PMsgResp1 = clientA.sendMessage("Shuai Hu", "Zhi Xu", "SB");
+            assertEquals(true, p2PMsgResp1.getSuccess());
+            assertEquals(false, p2PMsgResp1.getIsDelivered());
+
+            try (MongoCursor<Document> cursor = mongoDB.getCollection(MyDatabase.MESSAGES).find(eq(Message.FROM_USER, "Shuai Hu")).iterator()) {
+                if (cursor.hasNext()) {
```
Comment (hypnotichen) [inline src/test/com/Elessar/app/client/MyClientIntegTest.java]: get rid of this "if" and also verify the total number of documents returned are what we expect

File: src/test/com/Elessar/app/client/MyClientIntegTest.java [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+        final String serverURL = new StringBuilder().append("http://")
+                                                    .append(serverAddress).append(":")
+                                                    .append(serverPort).toString();
+        clientA_Port = 4000;
+        clientB_Port = 5000;
+        final MetricManager serverMetricManager = new MetricManager("ServerMetric", 1000000);
+        final MetricManager clientMetricManager = new MetricManager("ClientMetric", 1000000);
+
+        // Setup server
+        mongoDB = MongoClients.create("mongodb://localhost:27017").getDatabase("test");
```
Comment (hypnotichen) [inline src/test/com/Elessar/app/client/MyClientIntegTest.java]: better to use, say "MyClientIntegTest" as the database name so we won't accidentally mess up other tests' databases/collections

File: src/test/com/Elessar/app/client/MyClientIntegTest.java [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+                                                                                                   eq(Message.FROM_USER, "Zhi Xu"))).iterator()) {
+                assertEquals(false, cursor.hasNext());
+            }
+
+            LogonResponse logonResp1 = clientA.logOn("Shuai Hu", "4238902", clientA_Port);;
+            assertEquals(true, logonResp1.getSuccess());
+
+            P2PMsgResponse p2PMsgResp1 = clientA.sendMessage("Shuai Hu", "Zhi Xu", "SB");
+            assertEquals(true, p2PMsgResp1.getSuccess());
+            assertEquals(false, p2PMsgResp1.getIsDelivered());
```
Comment (hypnotichen) [inline src/test/com/Elessar/app/client/MyClientIntegTest.java]: We can also add another test here to verify server sends the undelivered messages to a user when that user logs on.
Basically:
- send a message to a offline user
- check the message is not delivered
- user logs on
- test sleeps for say 2 seconds
- the message should have been updated to "delivered" in mongo

File: src/test/com/Elessar/app/client/MyClientIntegTest.java [hunk 10]
```diff
@@ -0,0 +1,291 @@
...
+    }
+
+    @AfterClass
+    public static void tearDownAfterClass() throws Exception {
+        mongoDB.drop();
+    }
+
+
+    private void registerTestSetup() throws Exception {
+        mongoDB.getCollection(MyDatabase.USERS).deleteMany(new Document());
```
Comment (hypnotichen) [inline src/test/com/Elessar/app/client/MyClientIntegTest.java]: minor but probably should just drop the collection... easier to understand

**Comment by DIBL** [comment]
Then create index before every test?

**Pull request merged**